libr/bin/dwarf.c

```diff
@@ -213,98 +213,29 @@ static void r_bin_dwarf_header_fini(RBinDwarfLNPHeader *hdr) {
 		free (hdr->file_names);
 	}
 }
-
-static const ut8 *r_bin_dwarf_parse_lnp_header(
-	RBinFile *bf, const ut8 *buf, const ut8 *buf_end,
-	RBinDwarfLNPHeader *hdr, FILE *f, int mode) {
-	int i;
-	Sdb *s;
+// Parses source file header of DWARF version <= 4
+static const ut8 *parse_line_header_source(RBinFile *bf, const ut8 *buf, const ut8 *buf_end,
+	RBinDwarfLineHeader *hdr, FILE *f, Sdb *sdb, int mode) {
+	int i = 0;
 	size_t count;
 	const ut8 *tmp_buf = NULL;
 
-	if (!hdr || !bf || !buf) {
-		return NULL;
-	}
-
-	hdr->unit_length.part1 = READ32 (buf);
-	if (hdr->unit_length.part1 == DWARF_INIT_LEN_64) {
-		hdr->unit_length.part2 = READ64 (buf);
-	}
-
-	s = sdb_new (NULL, NULL, 0);
-
-	hdr->version = READ16 (buf);
-
-	if (hdr->unit_length.part1 == DWARF_INIT_LEN_64) {
-		hdr->header_length = READ64 (buf);
-	} else {
-		hdr->header_length = READ32 (buf);
-	}
-
-	if (buf_end-buf < 8) {
-		sdb_free (s);
-		return NULL;
-	}
-	hdr->min_inst_len = READ8 (buf);
-	if (hdr->version >= 4) {
-		hdr->max_ops_per_inst = READ8 (buf);
-	}
-	hdr->file_names = NULL;
-	hdr->default_is_stmt = READ8 (buf);
-	hdr->line_base = READ (buf, int8_t);
-	hdr->line_range = READ8 (buf);
-	hdr->opcode_base = READ8 (buf);
-
 	if (f) {
-		fprintf (f, " Header information:\n");
-		fprintf (f, "  Length:                             %d\n", hdr->unit_length.part1);
-		fprintf (f, "  DWARF Version:                      %d\n", hdr->version);
-		fprintf (f, "  Header Length:                      %"PFMT64d"\n", hdr->header_length);
-		fprintf (f, "  Minimum Instruction Length:         %d\n", hdr->min_inst_len);
-		fprintf (f, "  Maximum Operations per Instruction: %d\n", hdr->max_ops_per_inst);
-		fprintf (f, "  Initial value of 'is_stmt':         %d\n", hdr->default_is_stmt);
-		fprintf (f, "  Line Base:                          %d\n", hdr->line_base);
-		fprintf (f, "  Line Range:                         %d\n", hdr->line_range);
-		fprintf (f, "  Opcode Base:                        %d\n", hdr->opcode_base);
-		fprintf (f, "\n");
+		fprintf (f, " The Directory Table:\n");
 	}
-
-	if (hdr->opcode_base>0) {
-		hdr->std_opcode_lengths = calloc(sizeof(ut8), hdr->opcode_base);
-
-		if (f) {
-			fprintf (f, " Opcodes:\n");
-		}
-		for (i = 1; i <= hdr->opcode_base - 1; i++) {
-			if (buf + 2 > buf_end) {
-				break;
-			}
-			hdr->std_opcode_lengths[i] = READ (buf, ut8);
-			if (f) {
-				fprintf (f, "  Opcode %d has %d arg\n", i, hdr->std_opcode_lengths[i]);
-			}
-		}
-		if (f) {
-			fprintf (f, "\n");
-		}
-	} else {
-		hdr->std_opcode_lengths = NULL;
-	}
-
-	i = 0;
-	while (buf+1 < buf_end) {
-		int maxlen = R_MIN ((size_t)(buf_end-buf)-1, 0xfff);
-		int len = r_str_nlen ((const char*)buf, maxlen);
+	while (buf + 1 < buf_end) {
+		int maxlen = R_MIN ((size_t) (buf_end - buf) - 1, 0xfff);
+		int len = r_str_nlen ((const char *)buf, maxlen);
 		char *str = r_str_ndup ((const char *)buf, len);
-		if (len<1 || len >= 0xfff) {
+		if (len < 1 || len >= 0xfff) {
```

Comment:
0xfff seems like an arbitrary limit

Reply:
Yeah, but I guess it it's sufficient for any real dir length? And it's somewhat of an obscure protection against malformed DWARF. Not sure what the original writer intention was tho.